Sources/ComposableArchitecture/Store.swift

```diff
@@ -426,22 +411,57 @@ public final class Store<State, Action> {
       let effect = self.reducer(&currentState, action)
 
       var didComplete = false
+      let boxedTask = Box<Task<Void, Never>?>(wrappedValue: nil)
       let uuid = UUID()
-      let effectCancellable = effect.sink(
-        receiveCompletion: { [weak self] _ in
-          self?.threadCheck(status: .effectCompletion(action))
-          didComplete = true
-          self?.effectCancellables[uuid] = nil
-        },
-        receiveValue: { [weak self] effectAction in
-          self?.send(effectAction, originatingFrom: action, file: file, line: line, instrumentation: instrumentation)
-        }
-      )
+      let effectCancellable =
+        effect
+        .handleEvents(
+          receiveCancel: { [weak self] in
+            self?.threadCheck(status: .effectCompletion(action))
+            self?.effectCancellables[uuid] = nil
+          }
+        )
+        .sink(
+          receiveCompletion: { [weak self] _ in
+            self?.threadCheck(status: .effectCompletion(action))
+            boxedTask.wrappedValue?.cancel()
+            didComplete = true
+            self?.effectCancellables[uuid] = nil
+          },
+          receiveValue: { [weak self] effectAction in
+            guard let self = self else { return }
+            if let task = self.send(effectAction, originatingFrom: action, file: file, line: line, instrumentation: instrumentation) {
+              tasks.wrappedValue.append(task)
+            }
+          }
+        )
 
       if !didComplete {
+        let task = Task<Void, Never> { @MainActor in
+          for await _ in AsyncStream<Void>.never {}
```

Comment:
Does this simply create a task that literally does nothing ever so that the next line is not executed unless this parent task is canceled?

Any clues on why this form compared to the `withTaskCancellationHandler` like below?

Reply:
Had to try - devil's in the details :-)

You mean something like this?
```
        let task = Task<Void, Never> { @MainActor in
            await withTaskCancellationHandler { in
                effectCancellable.cancel()
            } operation: { @MainActor in
                for await _ in AsyncStream<Void>.never {}
            }
        }
```

It seems that the cancellation handler is this case can't just marked as `@MainActor`, so the `cancel` can be ran on a different context, which currently breaks tests.